REGRESSION: Emoji characters are sometimes drawn in text style erroneously

litherum opened the pull request:
#### 3f69e13f0317394affb24ab40d881b92f2f1299d
<pre>
REGRESSION: Emoji characters are sometimes drawn in text style erroneously
<a href="https://bugs.webkit.org/show_bug.cgi?id=259139">https://bugs.webkit.org/show_bug.cgi?id=259139</a>
rdar://111411175

Reviewed by Cameron McCormack.

In WebKit, we weren&apos;t treating emoji deliberately at all. We were treating them like any other
character - falling back to the first font in the fallback list that supports them. This is a
problem because there are some fonts (like STIXTwo) which support emoji code points, but
aren&apos;t colorful like emoji should be. In particular, on macOS, STIXTwo is earlier in the
platform&apos;s default font cascade list than Apple Color Emoji, so some emoji characters were being
rendered in text style, not emoji style.

The fix for this is to treat emoji deliberately. This patch adds two classification functions:
- One that classifies a character as either &quot;RequireText&quot;, &quot;RequireEmoji&quot; or &quot;NoPreference&quot;
- One that classifies a glyph in a font as either color or not

When we perform font fallback for a particular character, we can make sure that if the
character requires emoji style, that it is being rendered with a color glyph (and vice-versa
for characters that require text style - of which there are none right now, but the yet-to-
be-implemented `font-variant-emoji` property would create this situation). If the glyph from
the fallback font isn&apos;t compatible with the character, then the solution is to just keep
falling back to another font. If we get to the end of the list, we ask the platform to produce
a font for us, and we pass it a variation selector which indicates which kind of glyph/font
we need.

The first classification works by just asking Unicode (ICU) whether the character has the
Emoji_Presentation property. The second classification works by creating a bitmap of all the
glyphs at font creation time, which can then be queried at font fallback time.

(On the Linux ports, I couldn&apos;t figure out how to implement the second classification function.
See <a href="https://bugs.webkit.org/show_bug.cgi?id=259205.">https://bugs.webkit.org/show_bug.cgi?id=259205.</a> This means that the first classification
needs to report NoPreference for every character on those ports. We don&apos;t want a situation where
a character reports it requires emoji style, but we don&apos;t know how to ask a font if it has
colorful glyphs, so we assume no fonts have colorful glyphs, which means no fonts can render
the character.)

Because a particular glyph in a font is either text style or emoji style (and not both),
all the caches _inside_ Font don&apos;t need to be updated. The only caches which need to be updated
in this patch is the ones in FontCascadeFonts where different glyphs from different fonts are
combined together into a single data structure.

The test for this patch only tests the characters which Unicode says actually need emoji style.
All the other characters are up to the OS/UA&apos;s discretion about whether they should be emoji
or not. For example, iOS renders a bunch of more characters in emoji style than just what&apos;s
required by Unicode - but macOS doesn&apos;t. I also limited the set of characters from &quot;all the
characters Unicode says need emoji style&quot; to &quot;all the characters Unicode says need emoji style
that macOS renders as text style&quot; so the set isn&apos;t super big.

* LayoutTests/fast/text/emoji-style-expected.html: Added.
* LayoutTests/fast/text/emoji-style.html: Added.
* Source/WTF/wtf/BitVector.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/unicode/CharacterNames.h:
* Source/WebCore/PAL/pal/cf/CoreTextSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreTextSoftLink.h:
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::systemFallbackFontForCharacter const):
(WebCore::Font::glyphIsColorGlyph const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::glyphDataForCharacter const):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::FontCascadeFonts::glyphDataForSystemFallback):
(WebCore::FontCascadeFonts::glyphDataForVariant):
(WebCore::FontCascadeFonts::glyphDataForCharacter):
(WebCore::FontCascadeFonts::pruneSystemFallbacks):
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
* Source/WebCore/platform/graphics/GlyphPage.h:
(WebCore::GlyphData::GlyphData):
(WebCore::GlyphPage::glyphDataForIndex const):
(WebCore::GlyphPage::glyphForIndex const):
(WebCore::GlyphPage::isColorForIndex const):
(WebCore::GlyphPage::setGlyphForIndex):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.cpp:
(WebCore::SystemFallbackFontCache::systemFallbackFontForCharacter):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.h:
(WebCore::add):
(WebCore::CharacterFallbackMapKey::operator== const): Deleted.
* Source/WebCore/platform/graphics/cairo/FontCairo.cpp:
(WebCore::FontCascade::resolveEmojiPolicy):
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::FontCascade::resolveEmojiPolicy):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformInit):
* Source/WebCore/platform/graphics/coretext/GlyphPageCoreText.cpp:
(WebCore::GlyphPage::fill):
* Source/WebCore/platform/graphics/freetype/GlyphPageTreeNodeFreeType.cpp:
(WebCore::GlyphPage::fill):
* Source/WebCore/platform/graphics/opentype/OpenTypeVerticalData.cpp:
(WebCore::OpenTypeVerticalData::substituteWithVerticalGlyphs const):
* Source/WebCore/platform/graphics/win/GlyphPageTreeNodeCairoWin.cpp:
(WebCore::GlyphPage::fill):
* Source/WebCore/platform/text/TextFlags.h:

Canonical link: <a href="https://commits.webkit.org/266089@main">https://commits.webkit.org/266089@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd563a50a3e3034f19730a0565764b5d8b65f5c7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12859 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13185 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14599 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12278 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15687 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13206 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14976 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13025 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/13745 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10882 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15050 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11033 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11633 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18702 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/10937 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12108 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11802 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15004 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12171 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12263 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10166 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/12911 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11525 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11502 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3396 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3145 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15840 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13281 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12103 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3187 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/platform/graphics/cairo/FontCairo.cpp:
```diff
@@ -126,6 +127,30 @@ float Font::platformWidthForGlyph(Glyph glyph) const
+    case FontVariantEmoji::Emoji:
+        return ResolvedEmojiPolicy::RequireEmoji;
+    }
+}
```
Asking for help with this at https://webkit.slack.com/archives/CU65R5WQY/p1689189611828469

litherum commented on Source/WebCore/platform/graphics/SystemFallbackFontCache.cpp:
```diff
@@ -61,6 +62,22 @@ RefPtr<Font> SystemFallbackFontCache::systemFallbackFontForCharacter(const Font*
+        // FIXME: Is this the right place to add the variation selectors?
+        // Should this be done in platform-specific code instead?
+        // The fact that Core Text accepts this information in the form of variation selectors
+        // seems like a platform-specific quirk.
```
I should also state what the downside to doing it this suggested way is.

litherum replied:
https://github.com/WebKit/WebKit/pull/15811

litherum commented on Source/WebCore/platform/graphics/GlyphPage.h:
```diff
@@ -119,12 +130,14 @@ class GlyphPage : public RefCounted<GlyphPage> {
 
     const Font& m_font;
     Glyph m_glyphs[size] { };
+    BitVector m_isColor; // FIXME: This is too large for WTF::BitVector to keep the storage inline. Either use std::bitset or adjust WTF::BitVector.
```
This should migrate to `WTF::Bitmap`

litherum replied:
https://github.com/WebKit/WebKit/commit/da4d5fa288fc590dd6d77d598d66a61de11b6c2c

litherum commented on Source/WebCore/platform/graphics/FontCascadeFonts.h:
```diff
@@ -100,8 +101,8 @@ class FontCascadeFonts : public RefCounted<FontCascadeFonts> {
-    GlyphPageCacheEntry m_cachedPageZero;
-    HashMap<int, GlyphPageCacheEntry> m_cachedPages;
+    EnumeratedArray<ResolvedEmojiPolicy, GlyphPageCacheEntry, ResolvedEmojiPolicy::RequireEmoji> m_cachedPageZero;
+    EnumeratedArray<ResolvedEmojiPolicy, HashMap<int, GlyphPageCacheEntry>, ResolvedEmojiPolicy::RequireEmoji> m_cachedPages;
```
We should hook `EnumeratedArray` into `WTF::EnumTraits` so it can infer the max value

litherum replied:
https://github.com/WebKit/WebKit/pull/15809

litherum commented on Source/WebCore/platform/graphics/FontCascadeFonts.h:
```diff
@@ -100,8 +101,8 @@ class FontCascadeFonts : public RefCounted<FontCascadeFonts> {
 
-    GlyphPageCacheEntry m_cachedPageZero;
-    HashMap<int, GlyphPageCacheEntry> m_cachedPages;
+    EnumeratedArray<ResolvedEmojiPolicy, GlyphPageCacheEntry, ResolvedEmojiPolicy::RequireEmoji> m_cachedPageZero;
```
There is no reason to single out the zero page any more. It should just be part of the `HashMap`.

litherum commented on Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
```diff
@@ -399,8 +399,9 @@ static void opportunisticallyStartFontDataURLLoading(const FontCascadeDescriptio
-GlyphData FontCascadeFonts::glyphDataForVariant(UChar32 character, const FontCascadeDescription& description, FontVariant variant, unsigned fallbackIndex)
+GlyphData FontCascadeFonts::glyphDataForVariant(UChar32 character, const FontCascadeDescription& description, FontVariant variant, ResolvedEmojiPolicy resolvedEmojiPolicy, unsigned fallbackIndex)
 {
+    UNUSED_PARAM(resolvedEmojiPolicy);
```
Delete this.

litherum commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -330,7 +330,9 @@ GlyphData FontCascade::glyphDataForCharacter(UChar32 c, bool mirror, FontVariant
         c = u_charMirror(c);
 
-    return m_fonts->glyphDataForCharacter(c, m_fontDescription, variant);
+    auto emojiPolicy = resolveEmojiPolicy(FontVariantEmoji::Normal, c);
```
Is this going to be fast enough to do with no caching?

litherum replied:
Well, Core Text does this without any caching, and it's not super obvious if a hash lookup is significantly faster than a property lookup, soooooo probably yes

litherum commented on Source/WebCore/platform/graphics/coretext/GlyphPageCoreText.cpp:
```diff
@@ -65,7 +65,7 @@ bool GlyphPage::fill(UChar* buffer, unsigned bufferLength)
         auto theGlyph = glyphs[i * glyphStep];
         if (theGlyph && theGlyph != deletedGlyph) {
-            setGlyphForIndex(i, theGlyph);
+            setGlyphForIndex(i, theGlyph, false); // FIXME: Litherum Implement this.
```
Also the other ports' implementations of this function needs to be updated too.

heycam commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -226,6 +225,8 @@ class Font : public RefCounted<Font> {
     };
     const Attributes& attributes() const { return m_attributes; }
 
+    bool glyphIsColorGlyph(Glyph) const;
```
Should we name this `isColorGlyph` for consistency with `isZeroWidthSpaceGlyph`?

litherum replied:
Good idea.

heycam commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -221,6 +221,25 @@ void Font::platformInit()
+        } else
+#endif
+        {
+            m_emojiType = AllEmojiGlyphs { };
```
It's not clear to me what the result of using AllEmojiGlyphs is here, when CTFontCopyColorGlyphCoverage is not available (like on downlevels). Maybe no change for current behavior, since we don't implement font-variant-emoji (and so won't use ResolveEmojiPolicy::RequireText)?

litherum replied:
The idea is, unless you have glyph-by-glyph information, the font itself has to be classified as either a "color font" or a "non-color font." Color fonts are able to honor requests for emoji-style characters. So we say, if `CTFontGetSymbolicTraits(m_platformData.font()) & kCTFontTraitColorGlyphs` is true, act as if every glyph in the font is an emoji-style glyph, because the font is a color font.

heycam commented on Source/WebCore/platform/graphics/GlyphPage.h:
```diff
@@ -92,20 +95,28 @@ class GlyphPage : public RefCounted<GlyphPage> {
+
     // FIXME: Pages are immutable after initialization. This should be private.
-    void setGlyphForIndex(unsigned index, Glyph glyph)
+    void setGlyphForIndex(unsigned index, Glyph glyph, bool isColor)
```
Not a big fan of the `bool` argument, when there are call sites that are passing in `false` and it's not clear from reading what it means.

heycam reviewed (approved):
I did not verify the specific list of characters you are testing.

litherum committed:
REGRESSION: Emoji characters are sometimes drawn in text style erroneously
https://bugs.webkit.org/show_bug.cgi?id=259139
rdar://111411175

Reviewed by Cameron McCormack.

In WebKit, we weren't treating emoji deliberately at all. We were treating them like any other
character - falling back to the first font in the fallback list that supports them. This is a
problem because there are some fonts (like STIXTwo) which support emoji code points, but
aren't colorful like emoji should be. In particular, on macOS, STIXTwo is earlier in the
platform's default font cascade list than Apple Color Emoji, so some emoji characters were being
rendered in text style, not emoji style.

The fix for this is to treat emoji deliberately. This patch adds two classification functions:
- One that classifies a character as either "RequireText", "RequireEmoji" or "NoPreference"
- One that classifies a glyph in a font as either color or not

When we perform font fallback for a particular character, we can make sure that if the
character requires emoji style, that it is being rendered with a color glyph (and vice-versa
for characters that require text style - of which there are none right now, but the yet-to-
be-implemented `font-variant-emoji` property would create this situation). If the glyph from
the fallback font isn't compatible with the character, then the solution is to just keep
falling back to another font. If we get to the end of the list, we ask the platform to produce
a font for us, and we pass it a variation selector which indicates which kind of glyph/font
we need.

The first classification works by just asking Unicode (ICU) whether the character has the
Emoji_Presentation property. The second classification works by creating a bitmap of all the
glyphs at font creation time, which can then be queried at font fallback time.

(On the Linux ports, I couldn't figure out how to implement the second classification function.
See https://bugs.webkit.org/show_bug.cgi?id=259205. This means that the first classification
needs to report NoPreference for every character on those ports. We don't want a situation where
a character reports it requires emoji style, but we don't know how to ask a font if it has
colorful glyphs, so we assume no fonts have colorful glyphs, which means no fonts can render
the character.)

Because a particular glyph in a font is either text style or emoji style (and not both),
all the caches _inside_ Font don't need to be updated. The only caches which need to be updated
in this patch is the ones in FontCascadeFonts where different glyphs from different fonts are
combined together into a single data structure.

The test for this patch only tests the characters which Unicode says actually need emoji style.
All the other characters are up to the OS/UA's discretion about whether they should be emoji
or not. For example, iOS renders a bunch of more characters in emoji style than just what's
required by Unicode - but macOS doesn't. I also limited the set of characters from "all the
characters Unicode says need emoji style" to "all the characters Unicode says need emoji style
that macOS renders as text style" so the set isn't super big.

* LayoutTests/fast/text/emoji-style-expected.html: Added.
* LayoutTests/fast/text/emoji-style.html: Added.
* Source/WTF/wtf/BitVector.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/unicode/CharacterNames.h:
* Source/WebCore/PAL/pal/cf/CoreTextSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreTextSoftLink.h:
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::systemFallbackFontForCharacter const):
(WebCore::Font::glyphIsColorGlyph const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::glyphDataForCharacter const):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::FontCascadeFonts::glyphDataForSystemFallback):
(WebCore::FontCascadeFonts::glyphDataForVariant):
(WebCore::FontCascadeFonts::glyphDataForCharacter):
(WebCore::FontCascadeFonts::pruneSystemFallbacks):
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
* Source/WebCore/platform/graphics/GlyphPage.h:
(WebCore::GlyphData::GlyphData):
(WebCore::GlyphPage::glyphDataForIndex const):
(WebCore::GlyphPage::glyphForIndex const):
(WebCore::GlyphPage::isColorForIndex const):
(WebCore::GlyphPage::setGlyphForIndex):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.cpp:
(WebCore::SystemFallbackFontCache::systemFallbackFontForCharacter):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.h:
(WebCore::add):
(WebCore::CharacterFallbackMapKey::operator== const): Deleted.
* Source/WebCore/platform/graphics/cairo/FontCairo.cpp:
(WebCore::FontCascade::resolveEmojiPolicy):
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::FontCascade::resolveEmojiPolicy):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformInit):
* Source/WebCore/platform/graphics/coretext/GlyphPageCoreText.cpp:
(WebCore::GlyphPage::fill):
* Source/WebCore/platform/graphics/freetype/GlyphPageTreeNodeFreeType.cpp:
(WebCore::GlyphPage::fill):
* Source/WebCore/platform/graphics/opentype/OpenTypeVerticalData.cpp:
(WebCore::OpenTypeVerticalData::substituteWithVerticalGlyphs const):
* Source/WebCore/platform/graphics/win/GlyphPageTreeNodeCairoWin.cpp:
(WebCore::GlyphPage::fill):
* Source/WebCore/platform/text/TextFlags.h:

Canonical link: https://commits.webkit.org/266089@main
